29.md

```diff
@@ -6,153 +6,270 @@ Relay-based Groups
+Relays MAY choose to not to publish this information, restrict what pubkeys can fetch it, or reject any member lists published by users to the relay.
 
-# Rooms
+# Migration
 
-Rooms are defined by an arbitrary string containing any text. Names SHOULD be human-readable, and less than 30 characters.
+If a group needs to be moved from one relay to another, this can be done by publishing a `kind 9009`, with the following tags:
 
-Anyone MAY publish `kind xxxxx` index events containing `h` tags and the relay on which they exist. This is only a convenience for helping clients build room lists, and shouldn't be relied upon as being accurate or complete.  `h` tags MUST indicate what type the room is. Types of rooms and their corresponding event kinds are defined below.
+- `h` indicates the group id.
+- `r` with a `old` mark indicates the current relay.
+- `r` with a `new` mark indicates the relay the group is moving to.
 
 ```json
 {
-  "kind": "xxxxx",
+  "kind": "9009",
   "tags": [
-    ["h", "salamanders", "wss://relay.froglovers.com", "chat"],
-    ["h", "frogs", "wss://relay.froglovers.com", "chat"],
-    ["h", "frogs", "wss://relay.catlovers.com", "forum"]
-  ]
+    ["h", "<group-id>"],
+    ["r", "wss://relay.dead.com/", "old"],
+    ["r", "wss://relay.live.com/", "new"]
+  ],
 }
 ```
 
-Events posted to rooms MUST include a simple `h` tag matching the room of the form `["h", "<room-name>"]`. An empty string MAY be used to indicate the "global" namespace.
+These events SHOULD be published to the relay's current relay, and MAY be published more broadly if more durability is desired (for example in the case where the relay goes down permanently before the migration is published).
+
+Anyone MAY publish these events, but migrations not published by the current relay's pubkey should be treated as a relay fork, rather than a migration.
+
+# Moderation
+
+Any relay member MAY publish a `kind 1984` report to the relay (including a `-` tag if desired/supported). These reports MAY be used by the relay admin to delete events or ban users, or by clients to implement any moderation algorithm desired. The relay operator or moderators may also choose to escalate reports without banning content by publishing their own `kind 1984` event with the same tags. Reports need not include an `h` tag.
+
+# Federation
+
+By default, the same group id on different relays does not mean that the same content will exist in both places. However, relays MAY choose to actively federate with others by aggressively replicating content between the two, or cooperating in some other way.
+
+If a group is federated, this should be indicated on the `kind:39000` group metadata event using one or more `peer` tags indicating another relay url. The peer SHOULD NOT be considered valid unless the designation is mutual.
+
+Clients should take care when supporting simultaneous relay use, since missing context is possible if the federation is not implemented correctly.
```

Comment:
Maybe explicit mention relay hints for this particular situations.

Reply:
I address this a bit more at the very end of the document. Relay hints are ok for bringing in note parents, but that's only one kind of context. Stuff like chat rooms has implicit context that would be hard to directly link, even if we used the `previous` tag